File: services/crm-service/src/main/java/com/crm/crm_backend/model/entity/Quotation.java
package com.crm.crm_backend.model.entity;


import com.crm.crm_backend.model.enums.QuotationStatus;
import jakarta.persistence.*;
import lombok.*;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "quotations")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Quotation {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "quotation_number",
            nullable = false,
            unique = true)
    private String quotationNumber;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "customer_id")
    private Customer customer;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "opportunity_id")
    private Opportunity opportunity;

    @Column(name = "issue_date")
    private LocalDate issueDate;

    @Column(name = "expiry_date")
    private LocalDate expiryDate;

    @Column(precision = 15, scale = 2)
    private BigDecimal subtotal;

    @Column(precision = 15, scale = 2)
    private BigDecimal discount;

    @Column(precision = 15, scale = 2)
    private BigDecimal tax;

    @Column(name = "total_amount",
            precision = 15,
            scale = 2)
    private BigDecimal totalAmount;

    @Enumerated(EnumType.STRING)
    private QuotationStatus status;

    @Column(length = 2000)
    private String notes;

    private Boolean active;

    @OneToMany(
            mappedBy = "quotation",
            cascade = CascadeType.ALL,
            orphanRemoval = true
    )
    @Builder.Default
    private List<QuotationItem> items = new ArrayList<>();

    private LocalDateTime createdAt;

    private LocalDateTime updatedAt;

    @PrePersist
    public void prePersist() {

        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();

        if (status == null) {
            status = QuotationStatus.DRAFT;
        }

        if (active == null) {
            active = true;
        }
    }

    @PreUpdate
    public void preUpdate() {
        updatedAt = LocalDateTime.now();
    }
}

File: frontend/src/features/crm/types/quotation.ts
export type QuotationStatus =
  | "DRAFT"
  | "SENT"
  | "ACCEPTED"
  | "REJECTED"
  | "EXPIRED"
  | "CANCELLED";

export type QuotationItem = {
  id?: number;
  itemName: string | null;
  sku: string | null;
  description: string | null;
  quantity: number | null;
  unitPrice: number | null;
  totalPrice?: number | null;
};

export type Quotation = {
  id: number;
  quotationNumber: string;
  customerId: number;
  opportunityId: number;
  issueDate: string;
  expiryDate: string;
  subtotal: number | null;
  discount: number | null;
  tax: number | null;
  totalAmount: number | null;
  status: QuotationStatus;
  notes: string | null;
  active: boolean | null;
  salesOrderId: number | null;
  items: QuotationItem[] | null;
  createdAt: string | null;
  updatedAt: string | null;
};

export type QuotationCreateInput = {
  customerId: number;
  opportunityId: number;
  issueDate: string;
  expiryDate: string;
  notes?: string;
  applyPricingRules?: boolean;
  items: Array<{
    itemName: string;
    sku?: string;
    description?: string;
    quantity: number;
    unitPrice: number;
  }>;
};
